File: 2021/10/26/BitwiseComplement.java
/**
 * https://leetcode.com/problems/complement-of-base-10-integer/
 * @author Andrew Jarombek
 * @since 10/26/2021
 */

public class BitwiseComplement {
    public int bitwiseComplement(int n) {
        if (n == 0) {
            return 1;
        }

        int mask = 1;
        int m = n;

        while (m > 0) {
            mask *= 2;
            m /= 2;
        }

        return (mask - 1) ^ n;
    }
}

File: 2022/06/03/bitwiseComplement.ts
/**
 * https://leetcode.com/problems/complement-of-base-10-integer/
 * @author Andrew Jarombek
 * @since 6/3/2022
 */

function bitwiseComplement(n: number): number {
    if (n == 0)
        return 1;

    let result = 0;
    let i = 1;

    while (n > 0) {
        if (n % 2 == 0)
            result += i;

        i <<= 1;
        n >>= 1;
    }

    return result;
}